vdr/didsubject/did_document.go

```diff
@@ -0,0 +1,189 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package didsubject
+
+import (
+	"encoding/json"
+	"errors"
+	"github.com/google/uuid"
+	ssi "github.com/nuts-foundation/go-did"
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/nuts-node/jsonld"
+	"gorm.io/gorm"
+	"gorm.io/gorm/schema"
+	"time"
+)
+
+// DIDDocument is the gorm representation of the did_document_version table
+type DIDDocument struct {
+	ID        string `gorm:"primaryKey"`
+	DidID     string `gorm:"column:did"`
+	DID       DID    `gorm:"foreignKey:DidID;references:ID"`
+	CreatedAt int64  `gorm:"autoCreateTime:false"`
+	// UpdatedAt is the (unix) timestamp when the document was updated
+	// Also used to purge DID document changes that haven't been committed within a certain time frame
+	UpdatedAt           int64 `gorm:"autoUpdateTime:false"`
+	Version             int
+	VerificationMethods []VerificationMethod `gorm:"foreignKey:DIDDocumentID;references:ID"`
+	Services            []SqlService         `gorm:"foreignKey:DIDDocumentID;references:ID"`
+	// Raw contains the DID Document as generated by the specific method, important for hashing.
+	Raw string
+}
+
+func (d DIDDocument) TableName() string {
+	return "did_document_version"
+}
+
+var _ DIDDocumentManager = (*SqlDIDDocumentManager)(nil)
+var _ schema.Tabler = (*DID)(nil)
+
+// DIDDocumentManager is the interface to change data for the did_document table
+type DIDDocumentManager interface {
+	// CreateOrUpdate adds a new version of a DID document, starts at 1
+	// If the DID does not exist yet, it will be created
+	// It adds all verification methods, services, alsoKnownAs to the DID document
+	// Not passing any verification methods will create an empty DID document, deactivation checking should be done by the caller
+	CreateOrUpdate(did DID, verificationMethods []VerificationMethod, services []SqlService) (*DIDDocument, error)
+	// Latest returns the latest version of a DID document
+	// if notAfter is given, it will return the latest version before that time
+	Latest(did did.DID, notAfter *time.Time) (*DIDDocument, error)
+}
+
+// SqlDIDDocumentManager is the implementation of the DIDDocumentManager interface
+type SqlDIDDocumentManager struct {
+	tx *gorm.DB
+}
+
+// NewDIDDocumentManager creates a new DIDDocumentManager for an open transaction
+func NewDIDDocumentManager(tx *gorm.DB) *SqlDIDDocumentManager {
+	return &SqlDIDDocumentManager{tx: tx}
+}
+
+func (s *SqlDIDDocumentManager) CreateOrUpdate(did DID, verificationMethods []VerificationMethod, services []SqlService) (*DIDDocument, error) {
```

Comment:
I think there are race conditions possible on the update

Reply:
how exactly? The `SqlDIDDocumentManager` contains a reference to a transaction, so transaction isolation is in place.

Reply:
you can start 2 simultaneous transactions on the same DID. Both will act on the same "latest" version. Unless this is solved in the code not part of this PR

Reply:
adding a uniqueness constraint on `version` would at least makes one transaction fail